tensorzero-internal/src/inference/providers/helpers.rs

```diff
@@ -12,20 +13,100 @@ pub fn inject_extra_body(
     config: Option<&ExtraBodyConfig>,
     body: &mut serde_json::Value,
 ) -> Result<(), Error> {
-    let Some(body_map) = body.as_object_mut() else {
+    if !body.is_object() {
         return Err(Error::new(ErrorDetails::Serialization {
             message: "Body is not a map".to_string(),
         }));
-    };
+    }
     if let Some(extra_body) = config.as_ref() {
-        for (key, value) in extra_body.data.iter() {
-            // We intentionally overwrite any existing keys in the body.
-            body_map.insert(key.to_string(), value.clone());
+        for replacement in &extra_body.data {
+            write_json_pointer_with_parent_creation(
+                body,
+                &replacement.pointer,
+                replacement.value.clone(),
+            )?;
         }
     }
     Ok(())
 }
 
+// Copied from serde_json (MIT-licensed): https://github.com/serde-rs/json/blob/400eaa977f1f0a1c9ad5e35d634ed2226bf1218c/src/value/mod.rs#L259
+fn parse_index(s: &str) -> Option<usize> {
+    if s.starts_with('+') || (s.starts_with('0') && s.len() != 1) {
+        return None;
+    }
+    s.parse().ok()
+}
+
+// Based on https://github.com/serde-rs/json/blob/400eaa977f1f0a1c9ad5e35d634ed2226bf1218c/src/value/mod.rs#L834
+fn write_json_pointer_with_parent_creation(
+    mut value: &mut serde_json::Value,
+    pointer: &str,
+    target_value: Value,
+) -> Result<(), Error> {
+    if pointer.is_empty() {
+        return Err(Error::new(ErrorDetails::ExtraBodyReplacement {
+            message: "Pointer cannot be empty".to_string(),
+            pointer: pointer.to_string(),
+        }));
+    }
+    if !pointer.starts_with('/') {
+        return Err(Error::new(ErrorDetails::ExtraBodyReplacement {
+            message: "Pointer must start with '/'".to_string(),
+            pointer: pointer.to_string(),
+        }));
+    }
+
+    if pointer.ends_with('/') {
+        return Err(Error::new(ErrorDetails::ExtraBodyReplacement {
+            message: "Pointer cannot end with '/'".to_string(),
+            pointer: pointer.to_string(),
+        }));
+    }
+
+    let components = pointer
+        .split('/')
+        .skip(1)
+        .map(|x| x.replace("~1", "/").replace("~0", "~"));
+    for token in components {
+        match value {
+            Value::Object(map) => match map.entry(token.clone()) {
+                Entry::Occupied(occupied) => value = occupied.into_mut(),
+                Entry::Vacant(vacant) => {
+                    if parse_index(&token).is_some() {
+                        return Err(Error::new(ErrorDetails::ExtraBodyReplacement {
+                        message: format!("Cannot create array (index {token}). This is an edge case - please report this to TensorZero"),
+                        pointer: pointer.to_string(),
+                        }));
+                    } else {
+                        value = vacant.insert(Value::Object(Map::new()));
+                    }
+                }
```

Comment:
can you add a comment explaining this logic? it won't be obvious to future readers of the code what's going on and why. Include the context on why we made this decision pls